File: core/src/net/natruid/jungle/utils/Point.java
package net.natruid.jungle.utils;

public class Point {
    public static final ImmutablePoint[] adjacent = new ImmutablePoint[]{
        new ImmutablePoint(new Point(1, 0)),
        new ImmutablePoint(new Point(0, 1)),
        new ImmutablePoint(new Point(-1, 0)),
        new ImmutablePoint(new Point(0, -1))
    };

    public static final ImmutablePoint[] diagonals = new ImmutablePoint[]{
        new ImmutablePoint(new Point(1, 1)),
        new ImmutablePoint(new Point(-1, 1)),
        new ImmutablePoint(new Point(-1, -1)),
        new ImmutablePoint(new Point(1, -1))
    };

    public int x;
    public int y;

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public Point() {
        this(0, 0);
    }

    public Point set(int x, int y) {
        this.x = x;
        this.y = y;
        return this;
    }

    public Point set(Point target) {
        x = target.x;
        y = target.y;
        return this;
    }

    @Override
    public String toString() {
        return String.format("(%d, %d)", x, y);
    }

    public Point plus(Point other) {
        this.x += other.x;
        this.y += other.y;
        return this;
    }

    public Point plus(ImmutablePoint other) {
        this.x += other.getX();
        this.y += other.getY();
        return this;
    }

    public Point minus(Point other) {
        this.x -= other.x;
        this.y -= other.y;
        return this;
    }

    public Point minus(ImmutablePoint other) {
        this.x -= other.getX();
        this.y -= other.getY();
        return this;
    }

    public Point times(Point other) {
        this.x *= other.x;
        this.y *= other.y;
        return this;
    }

    public Point times(ImmutablePoint other) {
        this.x *= other.getX();
        this.y *= other.getY();
        return this;
    }

    public Point times(int i) {
        this.x *= i;
        this.y *= i;
        return this;
    }

    public Point div(Point other) {
        this.x /= other.x;
        this.y /= other.y;
        return this;
    }

    public Point div(ImmutablePoint other) {
        this.x /= other.getX();
        this.y /= other.getY();
        return this;
    }
}

File: core/src/net/natruid/jungle/utils/Point.kt
package net.natruid.jungle.utils

data class Point(var x: Int, var y: Int) {
    companion object {
        private val adjacent = arrayOf(
            ImmutablePoint(Point(1, 0)),
            ImmutablePoint(Point(0, 1)),
            ImmutablePoint(Point(-1, 0)),
            ImmutablePoint(Point(0, -1))
        )
        private val diagonals = arrayOf(
            ImmutablePoint(Point(1, 1)),
            ImmutablePoint(Point(-1, 1)),
            ImmutablePoint(Point(-1, -1)),
            ImmutablePoint(Point(1, -1))
        )
    }

    constructor() : this(0, 0)

    fun set(x: Int, y: Int) {
        this.x = x
        this.y = y
    }

    fun set(target: Point) {
        this.x = target.x
        this.y = target.y
    }

    override fun toString(): String {
        return "($x, $y)"
    }

    fun neighbor(index: Int, diagonal: Boolean = false): Point {
        return this + if (diagonal) diagonals[index] else adjacent[index]
    }

    fun neighbors(diagonal: Boolean = false): Array<Point> {
        return Array(4) {
            neighbor(it, diagonal)
        }
    }

    operator fun plusAssign(other: Point) {
        this.x += other.x
        this.y += other.y
    }

    operator fun minusAssign(other: Point) {
        this.x -= other.x
        this.y -= other.y
    }

    operator fun timesAssign(other: Point) {
        this.x *= other.x
        this.y *= other.y
    }

    operator fun divAssign(other: Point) {
        this.x /= other.x
        this.y /= other.y
    }

    operator fun plus(other: Point): Point {
        return Point(x + other.x, y + other.y)
    }

    operator fun minus(other: Point): Point {
        return Point(x - other.x, y - other.y)
    }

    operator fun times(other: Point): Point {
        return Point(x * other.x, y * other.y)
    }

    operator fun div(other: Point): Point {
        return Point(x / other.x, y / other.y)
    }

    operator fun plusAssign(other: ImmutablePoint) {
        this.x += other.x
        this.y += other.y
    }

    operator fun minusAssign(other: ImmutablePoint) {
        this.x -= other.x
        this.y -= other.y
    }

    operator fun timesAssign(other: ImmutablePoint) {
        this.x *= other.x
        this.y *= other.y
    }

    operator fun divAssign(other: ImmutablePoint) {
        this.x /= other.x
        this.y /= other.y
    }

    operator fun plus(other: ImmutablePoint): Point {
        return Point(x + other.x, y + other.y)
    }

    operator fun minus(other: ImmutablePoint): Point {
        return Point(x - other.x, y - other.y)
    }

    operator fun times(other: ImmutablePoint): Point {
        return Point(x * other.x, y * other.y)
    }

    operator fun div(other: ImmutablePoint): Point {
        return Point(x / other.x, y / other.y)
    }

    operator fun timesAssign(i: Int) {
        x *= i
        y *= i
    }
}
